File: 1_Review/julia/7_comprehensions.jl
# 7) Comprehensions (vector, set, dict)
# Run: julia 7_comprehensions.jl

numbers = [x for x in 0:9]
println("List of numbers:")
println(numbers)

even_numbers = [x for x in numbers if x % 2 == 0]
println("Even numbers:")
println(even_numbers)

odd_numbers = [x for x in numbers if x % 2 != 0]
println("Odd numbers:")
println(odd_numbers)

squared_numbers = [x^2 for x in numbers]
println("Squared numbers:")
println(squared_numbers)

double_num(x) = x * 2
numbers_list = [double_num(x) for x in numbers]
println(numbers_list)

numbers_set = Set([x for x in numbers])
println("Set of numbers:")
println(numbers_set)

dictionary_squared_numbers = Dict(string(x) => x^2 for x in numbers)
println("Dictionary of squared numbers:")
println(dictionary_squared_numbers)

File: 1_Review/javascript/7_comprehensions.js
// 7) Comprehensions (JS uses map/filter instead of Python comprehensions)
// Run: node 7_comprehensions.js

// Numbers 0 through 9
const numbers = Array.from({ length: 10 }, (_, i) => i);
console.log('List of numbers:');
console.log(numbers);

const evenNumbers = numbers.filter((x) => x % 2 === 0);
console.log('Even numbers:');
console.log(evenNumbers);

const oddNumbers = numbers.filter((x) => x % 2 !== 0);
console.log('Odd numbers:');
console.log(oddNumbers);

const squaredNumbers = numbers.map((x) => x ** 2);
console.log('Squared numbers:');
console.log(squaredNumbers);

function doubleNum(num) {
  return num * 2;
}

const numbersList = numbers.map(doubleNum);
console.log(numbersList);

const numbersSet = new Set(numbers);
console.log('Set of numbers:');
console.log(numbersSet);

// "Dictionary" of squared numbers (object)
const dictionarySquaredNumbers = Object.fromEntries(numbers.map((x) => [String(x), x ** 2]));
console.log('Dictionary of squared numbers:');
console.log(dictionarySquaredNumbers);
